mono/metadata/appdomain.c

```diff
@@ -706,56 +706,102 @@ mono_domain_has_type_resolve (MonoDomain *domain)
+		: mono_domain_try_type_resolve_builder (domain, (MonoReflectionTypeBuilder *)tb, &error);
 	mono_error_cleanup (&error);
-
 	return ret;
 }
 
-MonoReflectionAssembly *
-mono_domain_try_type_resolve_checked (MonoDomain *domain, char *name, MonoObject *tb, MonoError *error)
+/**
+ * mono_class_get_appdomain_do_type_resolve_method:
+ *
+ * This routine returns System.AppDomain.DoTypeResolve.
+ */
+static MonoMethod *
+mono_class_get_appdomain_do_type_resolve_method (void)
 {
 	static MonoMethod *method = NULL;
-	MonoReflectionAssembly *ret;
-	void *params [1];
 
-	error_init (error);
-
-	g_assert (domain != NULL && ((name != NULL) || (tb != NULL)));
+	if (method)
+		return method;
 
+	method = mono_class_get_method_from_name (mono_class_get_appdomain_class (), "DoTypeResolve", -1);
 	if (method == NULL) {
-		method = mono_class_get_method_from_name (mono_class_get_appdomain_class (), "DoTypeResolve", -1);
-		if (method == NULL) {
-			g_warning ("Method AppDomain.DoTypeResolve not found.\n");
-			return NULL;
-		}
+		g_warning ("Method AppDomain.DoTypeResolve not found.\n");
 	}
 
-	if (name) {
-		*params = (MonoObject*)mono_string_new_checked (mono_domain_get (), name, error);
-		return_val_if_nok (error, NULL);
-	} else
-		*params = tb;
+	return method;
+}
+
+/**
+ * mono_domain_try_type_resolve_name:
+ * \param domain application domain in which to resolve the type
+ * \param name the name of the type to resolve.
+ *
+ * This routine invokes the internal \c System.AppDomain.DoTypeResolve and returns
+ * the assembly that matches name.
+ *
+ * \returns A \c MonoReflectionAssembly or NULL if not found
+ */
+MonoReflectionAssembly *
+mono_domain_try_type_resolve_name (MonoDomain *domain, const char *name, MonoError *error)
+{
+	g_assert (domain);
+	g_assert (name);
+	g_assert (error);
 
-	ret = (MonoReflectionAssembly *) mono_runtime_invoke_checked (method, domain->domain, params, error);
+	error_init (error);
+
+	void *params [] = { mono_string_new_checked (mono_domain_get (), name, error) };
 	return_val_if_nok (error, NULL);
 
-	return ret;
+	MonoMethod * const method = mono_class_get_appdomain_do_type_resolve_method ();
+	MonoObject * const ret = mono_runtime_invoke_checked (method, domain->domain, params, error);
+	return_val_if_nok (error, NULL);
+
+	return (MonoReflectionAssembly *)ret;
+}
+
+/**
+ * mono_domain_try_type_resolve_builder:
+ * \param domain application domain in which to resolve the type
+ * \param tb A \c System.Reflection.Emit.TypeBuilder; tb.FullName is the name of the type to resolve.
+ *
+ * This routine invokes the internal \c System.AppDomain.DoTypeResolve and returns
+ * the assembly that matches tb.FullName.
+ *
+ * \returns A \c MonoReflectionAssembly or NULL if not found
+ */
+MonoReflectionAssembly *
+mono_domain_try_type_resolve_builder (MonoDomain *domain, MonoReflectionTypeBuilder *tb, MonoError *error)
+{
+	g_assert (domain);
+	g_assert (tb);
+	g_assert (error);
+
+	error_init (error);
```

Comment:
The function isn't a leaf function nor set the error, so no need for that `error_init`.

Reply:
I think I was following preexisting pattern, but I agree the code is dodgy here and elsewhere. Howabout we initialize locals? That looks about right, will do.